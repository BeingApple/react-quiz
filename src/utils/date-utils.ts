import moment from "moment"
import { Moment, unitOfTime } from "moment"

type Unit = {
  unit: unitOfTime.Base
  string: string
}

const units: Array<Unit> = [
  {unit: "years", string: "년"},
  {unit: "months", string: "개월"}, 
  {unit: "days", string: "일"}, 
  {unit: "hours", string: "시간"}, 
  {unit: "minutes", string: "분"},
  {unit: "seconds", string: "초"},
]

export const durationToString = (duration: moment.Duration) => {
  let string = ""

  for (const unit of units) {
    const unitTime = duration.get(unit.unit)
    if (unitTime > 0) {
      string += unitTime + unit.string + " "
    }
  }

  return string.trim();
}

export const dateToString = (date?: Moment) => {
  date = (typeof date === 'string' || date instanceof String) ? moment(date) : date

  return date?.format("yyyy-MM-DD HH:mm:ss") ?? ""
}