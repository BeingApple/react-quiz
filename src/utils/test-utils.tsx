import React, { PropsWithChildren } from 'react'
import { render } from '@testing-library/react'
import type { RenderOptions } from '@testing-library/react'
import { Provider as ReduxProvider } from 'react-redux'

import { setupStore, type AppStore, type RootState } from '@/store/store'
// As a basic setup, import your same slice reducers
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>
  store?: AppStore
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  const queryClient = new QueryClient()

  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return (
      <ReduxProvider store={store}>
        <QueryClientProvider client={queryClient}>
          {children}
        </QueryClientProvider>
      </ReduxProvider>
    )
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}