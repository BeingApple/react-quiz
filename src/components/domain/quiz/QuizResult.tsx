import QuizPaper from "../../ui-components/QuizPaper"
import { Box, Button, Card, CardContent, Divider, Stack, Typography } from "@mui/material"
import { Moment } from "moment"
import { Quiz } from "@/types/quiz-types"
import { useEffect, useState } from "react"
import moment from "moment"
import { durationToString } from "@/utils/date-utils"
import { Chart as ChartJS, ArcElement, Tooltip, Legend, ChartData } from "chart.js";
import { Pie } from "react-chartjs-2";

ChartJS.register(ArcElement, Tooltip, Legend);

type Props = {
  startAt: Moment | null
  endAt: Moment | null
  correctList: Array<Quiz>
  wrongList: Array<Quiz>
  restart: () => void 
}

const QuizResult = ({startAt, endAt, correctList, wrongList, restart}: Props) => {
  const [timeString, setTimeString] = useState('')
  const [chartData, setChartData] = useState<ChartData<"pie", number[], unknown>>()

  useEffect(() => {
    if (startAt && endAt) {
      const duration = moment.duration(endAt.diff(startAt))
      setTimeString(durationToString(duration))
    }
  }, [startAt, endAt])

  useEffect(() => {
    const data = {
      labels: ['정답', '오답'],
      datasets: [{
        label: '갯수',
        data: [correctList.length, wrongList.length],
        backgroundColor: [
          'rgba(75, 192, 192, 0.2)',
          'rgba(255, 99, 132, 0.2)',
        ],
        borderColor: [
          'rgba(75, 192, 192, 1)',
          'rgba(255, 99, 132, 1)'
        ],
        borderWidth: 1,
      }]
    }

    setChartData(data)
  }, [correctList, wrongList])

  return (
    <QuizPaper>
      <Box sx={{p: 2}}>
        <Typography variant="h5" component="div">결과</Typography>
      </Box>
      <Divider />
      <Box sx={{p: 2}}>
        <Stack spacing={2}>
          <QuizPaper>
            <Typography variant="body2" color="text.secondary">소요된 시간 : {timeString}</Typography>
          </QuizPaper>
          <QuizPaper>
            <Typography variant="body2" color="text.secondary">정답 개수 : {correctList.length}</Typography>
          </QuizPaper>
          <QuizPaper>
            <Typography variant="body2" color="text.secondary">오답 개수 : {wrongList.length}</Typography>
          </QuizPaper>
          <Card>
            <CardContent>
              {chartData && <Pie data={chartData} /> }
            </CardContent>
          </Card>
        </Stack>
      </Box>
      <Button onClick={restart}>재시작</Button>
    </QuizPaper>
  )
}

export default QuizResult