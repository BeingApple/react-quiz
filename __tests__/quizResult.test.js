import '@testing-library/jest-dom'
import QuizResult from '@/components/domain/quiz/result'
import { durationToString } from "@/utils/date-utils"
import { renderWithProviders } from '../src/utils/test-utils'
import moment from 'moment';

describe('<QuizResult />', () => {
  const sampleItem = {
    type: 'multiple',
    difficulty: 'medium',
    category: 'TEST CATEGORY',
    question: 'TEST QUESTION',
    correct_answer: 'CORRENT ANSWER',
    incorrect_answers: ['INCORRECT ANSWER 1', 'INCORRECT ANSWER 2', 'INCORRECT ANSWER 3'],
    answers: ['CORRENT ANSWER', 'INCORRECT ANSWER 1', 'INCORRECT ANSWER 2', 'INCORRECT ANSWER 3'],
  }

  const sampleResult = {
    startAt: moment('2024-02-19 00:00:00'),
    endAt: moment('2024-02-19 00:01:24'),
    correctList: [sampleItem, sampleItem, sampleItem],
    wrongList: [sampleItem, sampleItem]
  }

  const setup = (props = {}) => {
    const initialProps = sampleResult
    const utils = renderWithProviders(<QuizResult {...initialProps} {...props} />)

    const { getByText } = utils

    const startAt = props.startAt || initialProps.startAt
    const endAt = props.endAt || initialProps.endAt
    const duration = moment.duration(endAt.diff(startAt))
    const timeString = durationToString(duration)
    const correctCount = props.correctList?.length || initialProps.correctList?.length
    const wrongCount = props.wrongList?.length || initialProps.wrongList?.length

    const time = getByText(`소요된 시간 : ${timeString}`)
    const correctCountComponent = getByText(`정답 개수 : ${correctCount}`)
    const wrongCountComponent = getByText(`오답 개수 : ${wrongCount}`)

    return {
      ...utils,
      time,
      correctCountComponent,
      wrongCountComponent
    };
  };

  it('matches snapshot', () => {
    const { container } = setup()
    expect(container).toMatchSnapshot();
  });

  it('has time spent, correct and incorrect count', () => {
    const { time, correctCountComponent, wrongCountComponent } = setup()
    expect(time).toBeTruthy()
    expect(correctCountComponent).toBeTruthy()
    expect(wrongCountComponent).toBeTruthy()
  });
});