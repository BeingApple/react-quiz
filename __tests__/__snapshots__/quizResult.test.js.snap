// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QuizResult /> matches snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-rccqv0-MuiPaper-root"
  >
    <div
      class="MuiBox-root css-19midj6"
    >
      <div
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        결과
      </div>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
    <div
      class="MuiBox-root css-19midj6"
    >
      <div
        class="MuiStack-root css-1sazv7p-MuiStack-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-rccqv0-MuiPaper-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            소요된 시간 : 
            1분 24초
          </p>
        </div>
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-rccqv0-MuiPaper-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            정답 개수 : 
            3
          </p>
        </div>
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-rccqv0-MuiPaper-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            오답 개수 : 
            2
          </p>
        </div>
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          />
        </div>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      재시작
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
